nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,194 @@
+              type = types.str;
+              default = "::1";
+              description = "Address to listen on for connections by the reverse proxy/tls terminator.";
+            };
+            global.database_path = mkOption {
+              type = types.str;
+              default = "/var/lib/matrix-conduit/";
+              readOnly = true;
+              description = ''
+                Path to the conduit database, the directory where conduit will save its data.
+                Note that due to using the DynamicUser feature of systemd, this value should not be changed
+                and is set to be read only.
+              '';
+            };
+          };
+        };
+        default = {};
+        description = ''
+            Generates the conduit.toml configuration file. Refer to
+            <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+            for details on supported values.
+            Note that database_path can not be edited because the service's reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = !(hasInfix ":" cfg.settings.global.server_name && cfg.nginx.enable);
+          message = ''
+            It appears you are trying to specify a port in the server_name variable.
+            This is not supported by the conduit nginx module. Please define your own vhost.
+          '';
+        }
+      ];
+
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = lib.mkMerge ([
+          { CONDUIT_CONFIG = configFile; }
+          cfg.extraEnvironment
+        ]);
+        serviceConfig = {
+          DynamicUser = true;
+          User = "conduit";
+          LockPersonality = true;
+          MemoryDenyWriteExecute = true;
+          ProtectClock = true;
+          ProtectControlGroups = true;
+          ProtectHostname = true;
+          ProtectKernelLogs = true;
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          PrivateDevices = true;
+          PrivateMounts = true;
+          PrivateUsers = true;
+          RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = [
+            "@system-service"
+            "~@privileged"
+          ];
+          StateDirectory = "matrix-conduit";
+          ExecStart = "${cfg.package}/bin/conduit";
+          Restart = "on-failure";
+          RestartSec = 10;
+          StartLimitBurst = 5;
+        };
+      };
+      services.nginx = mkIf cfg.nginx.enable {
+        enable = mkDefault true;
+        virtualHosts.${cfg.settings.global.server_name} = {
```

Comment:
Please think about situation where conduit is responsible for handling `example.com` but the actual conduit instance is on `asdfxcvbasfdasf.somewhere.org`. This is a very common setup for everyone that runs their website on another server / service.

I currently run two home servers that way:
`kack.it` is the name of the server but it is served by `matrix.kack.it`. Same with `nixos.dev` (which is the server) but is served by `matrix.nixos.dev`.

Reply:
See also https://github.com/NixOS/nixpkgs/pull/124435#discussion_r641516140

Reply:
I don't see why this shouldn't be handled by disabling the nginx support in the module and adding your own proxy. In nearly all more complex situations this will be a requirement anyway and the nginx part has already been criticised for being feature creep, so I'm hesitant of adding even more features to it.

Reply:
This isn't a complex thing. It should be the default. The other case should be the non-standard case.

Reply:
This seems to be the last blocker on this PR, and it may take some work to resolve. How about splitting it up into two pull requests – the second one containing the nginx option –, so that we can finally merge the most important part?

Reply:
Another possibilty would be to merge this as-is, since alle the features work as intended and let people that want additional features do a separate PR for them.

Reply:
If the currently requested change can be introduced without breaking backwards compatibility, that's fine for me.